# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/websocket/WebSocketChannel.h
# Commit: 432c7e9dc19f
# Full Hash: 432c7e9dc19fffeb05f470f6eb7bc9f619c82a2d
# Author: Kershaw Chang <kershaw@mozilla.com>
# Date: 2020-08-05 21:44:54
# Regressor Bug: 1497249
# File Overlap Count: 4
# Description:
#   Bug 1497249 - P1: Introduce nsWebSocketConnection interface for separating socket manipulation r=michal,necko-reviewers
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D30023
# ==============================================================================

diff -r 0d22ad297b19 -r 432c7e9dc19f netwerk/protocol/websocket/WebSocketChannel.h
--- a/netwerk/protocol/websocket/WebSocketChannel.h	Wed Aug 05 16:44:36 2020 +0000
+++ b/netwerk/protocol/websocket/WebSocketChannel.h	Mon Aug 03 10:03:08 2020 +0000
@@ -19,6 +19,7 @@
 #include "nsIProtocolProxyCallback.h"
 #include "nsIChannelEventSink.h"
 #include "nsIHttpChannelInternal.h"
+#include "nsIWebSocketConnection.h"
 #include "BaseWebSocketChannel.h"
 
 #include "nsCOMPtr.h"
@@ -63,15 +64,14 @@
 class WebSocketChannel : public BaseWebSocketChannel,
                          public nsIHttpUpgradeListener,
                          public nsIStreamListener,
-                         public nsIInputStreamCallback,
-                         public nsIOutputStreamCallback,
                          public nsITimerCallback,
                          public nsIDNSListener,
                          public nsIObserver,
                          public nsIProtocolProxyCallback,
                          public nsIInterfaceRequestor,
                          public nsIChannelEventSink,
-                         public nsINamed {
+                         public nsINamed,
+                         public nsIWebSocketConnectionListener {
   friend class WebSocketFrame;
 
  public:
@@ -79,8 +79,6 @@
   NS_DECL_NSIHTTPUPGRADELISTENER
   NS_DECL_NSIREQUESTOBSERVER
   NS_DECL_NSISTREAMLISTENER
-  NS_DECL_NSIINPUTSTREAMCALLBACK
-  NS_DECL_NSIOUTPUTSTREAMCALLBACK
   NS_DECL_NSITIMERCALLBACK
   NS_DECL_NSIDNSLISTENER
   NS_DECL_NSIPROTOCOLPROXYCALLBACK
@@ -88,6 +86,7 @@
   NS_DECL_NSICHANNELEVENTSINK
   NS_DECL_NSIOBSERVER
   NS_DECL_NSINAMED
+  NS_DECL_NSIWEBSOCKETCONNECTIONLISTENER
 
   // nsIWebSocketChannel methods BaseWebSocketChannel didn't implement for us
   //
@@ -142,6 +141,7 @@
 
   void EnqueueOutgoingMessage(nsDeque<OutboundMessage>& aQueue,
                               OutboundMessage* aMsg);
+  void DoEnqueueOutgoingMessage();
 
   void PrimeNewOutgoingMessage();
   void DeleteCurrentOutGoingMessage();
@@ -212,9 +212,7 @@
   nsCString mHost;
   nsString mEffectiveURL;
 
-  nsCOMPtr<nsISocketTransport> mTransport;
-  nsCOMPtr<nsIAsyncInputStream> mSocketIn;
-  nsCOMPtr<nsIAsyncOutputStream> mSocketOut;
+  nsCOMPtr<nsIWebSocketConnection> mConnection;
 
   nsCOMPtr<nsITimer> mCloseTimer;
   uint32_t mCloseTimeout; /* milliseconds */
@@ -280,17 +278,16 @@
   uint32_t mBuffered;
   uint32_t mBufferSize;
 
-  // These are for the send buffers
-  const static int32_t kCopyBreak = 1000;
-
   OutboundMessage* mCurrentOut;
-  uint32_t mCurrentOutSent;
   nsDeque<OutboundMessage> mOutgoingMessages;
   nsDeque<OutboundMessage> mOutgoingPingMessages;
   nsDeque<OutboundMessage> mOutgoingPongMessages;
-  uint32_t mHdrOutToSend;
+  uint32_t mHdrOutSize;
   uint8_t* mHdrOut;
-  uint8_t mOutHeader[kCopyBreak + 16];
+  // This is used to store the frame header and the close reason.
+  // Since the length of close reason can not be larger than 123, 256 is
+  // enough here.
+  uint8_t mOutHeader[256];
   UniquePtr<PMCECompression> mPMCECompressor;
   uint32_t mDynamicOutputSize;
   uint8_t* mDynamicOutput;